File: solutions/csharp/interest-is-interesting/1/InterestIsInteresting.cs
using System;

static class SavingsAccount
{
        public static float InterestRate(decimal balance)
        {
            while (balance < 0) return 3.213f;
            while (balance >= 0 && balance < 1000) return 0.5f;
            while (balance > 0 && balance >= 1000 && balance < 5000) return 1.621f;
            while (balance > 0 && balance >= 5000) return 2.475f;

            return 0;
        }

        public static decimal Interest(decimal balance)
        {
            while (balance <= 0) return ((decimal)SavingsAccount.InterestRate(balance)/100)*balance;
            while (balance > 0 && balance < 1000) return ((decimal)SavingsAccount.InterestRate(balance) / 100) * balance;
            while (balance > 0 && balance >= 1000 && balance < 5000) return ((decimal)SavingsAccount.InterestRate(balance) / 100) * balance;
            while (balance > 0 && balance >= 5000) return ((decimal)SavingsAccount.InterestRate(balance) / 100) * balance;
            return 0;
        }

        public static decimal AnnualBalanceUpdate(decimal balance)
        {
            while (balance <= 0) return (decimal)SavingsAccount.Interest(balance) + balance;
            while (balance > 0 && balance < 1000) return (decimal)SavingsAccount.Interest(balance) + balance;
            while (balance > 0 && balance >= 1000 && balance < 5000) return (decimal)SavingsAccount.Interest(balance) + balance;
            while (balance > 0 && balance >= 5000) return (decimal)SavingsAccount.Interest(balance) + balance;
            return 0;
        }

        public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
        {
        int years = 0;
        while (balance < targetBalance)
        {
            balance = AnnualBalanceUpdate(balance);
            years++;
        }
        return years;
        }
}

File: solutions/cpp/interest-is-interesting/1/interest_is_interesting.cpp
// interest_rate returns the interest rate for the provided balance.
double interest_rate(double balance) {
    // TODO: Implement the interest_rate function
    double interest = 0.0;
    if(balance < 0) interest = 3.213;        
    else if(balance >= 0 && balance < 1000) interest = 0.5;
    else if(balance >= 1000 && balance < 5000) interest = 1.621;
    else if(balance >= 5000) interest = 2.475;
    
    return interest;
}

// yearly_interest calculates the yearly interest for the provided balance.
double yearly_interest(double balance) {
    // TODO: Implement the yearly_interest function
    return (interest_rate(balance) / 100.0) * balance;
}

// annual_balance_update calculates the annual balance update, taking into
// account the interest rate.
double annual_balance_update(double balance) {
    // TODO: Implement the annual_balance_update function
    return yearly_interest(balance) + balance;
}

// years_until_desired_balance calculates the minimum number of years required
// to reach the desired balance.
int years_until_desired_balance(double balance, double target_balance) {
    // TODO: Implement the years_until_desired_balance function
    int year = 0;
    for (year = 0; balance < target_balance; year++) 
    {
        balance = annual_balance_update(balance);         
    }
    return year;
}
